CTCP-2105: Add EIS Connector and AccessCodeService

drnaylor opened the pull request:
This uses a best guess as to what we can do with EIS. This will need altering once we get the EIS interface documents.

drnaylor committed:
CTCP-2105: Add EIS Connector and AccessCodeService

drnaylor committed:
CTCP-2105: Add more tests

drnaylor committed:
CTCP-2105: Align with stub

liandra-giacomine commented on app/uk/gov/hmrc/ctcguaranteebalancerouter/config/EISInstanceConfig.scala:
```diff
@@ -0,0 +1,54 @@
+  retryConfig: RetryConfig
+) {
+
+  lazy val accessCodeUrl: String = s"$protocol://$host:$port${uris.accessCode}"
```
Do the accessCode and Balance uri change in diff environments? Just threw me off a bit because in other services I believe we keep these as constants

drnaylor replied:
I don't know what it's going to look like in EIS -- we do have it configurable in other services: https://github.com/hmrc/transit-movements-trader-eis-router/blob/2bccd04407c0526421591cb478b895fbd7f343af/conf/application.conf#L139

It needs to be configurable anyway, because I don't expect that EIS will include `ctc-guarantee-balance-eis-stub` in their URL. Additionally, it'll be a different URL for XI vs GB.

liandra-giacomine replied:
Oh yeah haha makes sense!

liandra-giacomine commented on app/uk/gov/hmrc/ctcguaranteebalancerouter/connectors/EISConnector.scala:
```diff
@@ -16,6 +16,122 @@
+          .setHeader(headerCarrier.headersForUrl(headerCarrierConfig)(eisInstanceConfig.accessCodeUrl): _*)
+          .execute[Either[UpstreamErrorResponse, HttpResponse]]
+          .map {
+            case Right(x) => Right(x.json)
```
Can we rename x to errorResponse and httpResponse or something like that? Or at least make the Right and Left value names different.

liandra-giacomine commented on app/uk/gov/hmrc/ctcguaranteebalancerouter/connectors/Retries.scala:
```diff
@@ -0,0 +1,72 @@
+      policy.decideNextRetry(status).map {
+        case r @ DelayAndRetry(delay) =>
+          if (System.currentTimeMillis() + delay.toMillis > endTime) GiveUp else r
+        case GiveUp => GiveUp
```
Can be reduced to:
```
    def decideNextRetry(status: RetryStatus): M[PolicyDecision] =
      policy.decideNextRetry(status).map {
        case r @ DelayAndRetry(delay) if System.currentTimeMillis() + delay.toMillis <= endTime => r
        case _                                                                                 => GiveUp
      }
```

drnaylor replied:
That's true. The only thing I'm a little nervous about is if the library is updated and another status appears, we'll get a warning then (because I think the status is `sealed`), having the wildcard means we won't necessarily realise that.

On the flip side, is that really likely to happen? Probably not. Just playing devil's advocate.

drnaylor replied:
I've gone for

```scala
{
        case r @ DelayAndRetry(delay) if System.currentTimeMillis() + delay.toMillis <= endTime => r
        case _: DelayAndRetry | GiveUp                                                          => GiveUp
      }
```

drnaylor committed:
CTCP-2105: Slight tidy up

drnaylor committed:
CTCP-2105: Add metrics

drnaylor commented:
I forgot to add metrics, so this needs another review.

drnaylor committed:
CTCP-2105: Ensure the V2 http client is available

drnaylor committed:
CTCP-2105: Improve the distinction between connector and service

I wasn't particularly happy with how I'd separated things and I felt like I'd blurred the lines a bit. The service doesn't really care about what format the data comes in, it just wants the data, nor should it have really dealt with HTTP status codes. I've confined that logic to the connector, the service now only gets an appropriate error or the deserialised object.

liandra-giacomine commented on app/uk/gov/hmrc/ctcguaranteebalancerouter/connectors/EISConnector.scala:
```diff
@@ -93,32 +99,38 @@ class EISConnectorImpl(
-            .map {
-              case Right(httpResponse)         => Right(httpResponse.json)
-              case Left(upstreamErrorResponse) => Left(RoutingError.Upstream(upstreamErrorResponse))
+            .map[Either[ConnectorError, AccessCodeResponse]] {
```
Can condense the code a bit by getting execute to deserialise AccessCodeResponse:
```

.execute[Either[UpstreamErrorResponse, AccessCodeResponse]]
            .map {
              case Right(accessCodeResponse) => Right(accessCodeResponse)
              // TODO: Improve once we know what EIS will do
              case Left(UpstreamErrorResponse(_, statusCode, _, _)) if statusCode == BAD_REQUEST || statusCode == NOT_FOUND => Left(ConnectorError.NotFound)
              case Left(upstreamErrorResponse)                                                                              => Left(ConnectorError.Upstream(upstreamErrorResponse))
            }
            .recover {
              case JsResultException(_) => Left(ConnectorError.FailedToDeserialise)
              case NonFatal(e) =>
                logger.error(s"Request Error: Routing to $code failed to retrieve data with message ${e.getMessage}. Correlation ID: $correlationId.")
                Left(ConnectorError.Unexpected("message", Some(e)))
            }

```

drnaylor replied:
I feel this is six of one and half a dozen of the other. I decided to do it this way primarily to avoid creating an exception (`JsResultException`) which can be expensive (generating stack traces and all), something we didn't need to do.

liandra-giacomine replied:
Fair enough!

liandra-giacomine commented on app/uk/gov/hmrc/ctcguaranteebalancerouter/services/AccessCodeService.scala:
```diff
@@ -51,25 +46,20 @@ class AccessCodeServiceImpl @Inject() (connectorProvider: EISConnectorProvider)
-    case x                                                                                      => AccessCodeError.Routing(x)
+  private def handleRoutingError(error: ConnectorError): AccessCodeError = error match {
+    case ConnectorError.NotFound                   => AccessCodeError.NotFound
+    case ConnectorError.FailedToDeserialise        => AccessCodeError.FailedToDeserialise
```
I think we'll need our ConvertError and PresentationError classes but that can be implemented when the controller is implemented.

drnaylor replied:
Yup, that's when they'll happen.

derekspalmer reviewed (approved):
LGTM.
Is the retry policy from requirements or a previous implementation?

drnaylor commented:
> Is the retry policy from requirements or a previous implementation?

More the latter, but MDTP talk about resilience in connections out -- this is one of the things they suggest we do, especially as we go to EIS.
